apollo-federation/src/sources/connect/json_selection/README.md

```diff
@@ -574,25 +637,77 @@ registered in the `JSONSelection` parser in order to be recognized.
-list->slice($args.start, $args.end)
-list->reverse
-some.value->times(2)
-some.value->plus($addend)
-some.value->minus(100)
-some.value->div($divisor)
-isDog: kind->eq("dog")
-isNotCat: kind->neq("cat")
-__typename: kind->match({ "dog": "Dog", "cat": "Cat" })
-decoded: utf8Bytes->decode("utf-8")
-utf8Bytes: string->encode("utf-8")
-encoded: bytes->encode("base64")
+# The ->echo method returns its first input argument as-is, ignoring
+# the input data. Useful for embedding literal values, as in
+# $->echo("give me this string"), or wrapping the input value.
+__typename: $->echo("Book")
+wrapped: field->echo({ fieldValue: @ })
+
+# Returns the type of the data as a string, e.g. "object", "array",
+# "string", "number", "boolean", or "null". Note that `typeof null` is
+# "object" in JavaScript but "null" for our purposes.
+typeOfValue: value->typeof
+
+# When invoked against an array, ->map evaluates its first argument
+# against each element of the array, binding the element values to `@`,
+# and returns an array of the results. When invoked against a non-array,
+# ->map evaluates its first argument against that value and returns the
+# result without wrapping it in an array.
+doubled: numbers->map(@->mul(2))
+types: values->map(@->typeof)
+
+# Returns true if the data is deeply equal to the first argument, false
+# otherwise. Equality is solely value-based (all JSON), no references.
+isObject: value->typeof->eq("object")
+
+# Takes any number of pairs [candidate, value], and returns value for
+# the first candidate that equals the input data. If none of the
+# pairs match, a runtime error is reported, but a single-element
+# [<default>] array as the final argument guarantees a default value.
+__typename: kind->match(
+    ["dog", "Canine"],
+    ["cat", "Feline"],
+    ["Exotic"]
+)
+
+# Like ->match, but expects the first element of each pair to evaluate
+# to a boolean, returning the second element of the first pair whose
+# first element is true. This makes providing a final catch-all case
+# easy, since the last pair can be [true, <default>].
+__typename: kind->matchIf(
+    [@->eq("dog"), "Canine"],
+    [@->eq("cat"), "Feline"],
+    [true, "Exotic"]
+)
+
+# Arithmetic methods, supporting both integers and floating point values,
+# similar to JavaScript.
+sum: $.a->add($.b)->add($.c)
+difference: $.a->sub($.b)->sub($.c)
+product: $.a->mul($.b, $.c)
+quotient: $.a->div($.b)
+remainder: $.a->mod($.b)
+
+# Array/string methods
+first: list->first
+last: list->last
+index3: list->get(3)
+secondToLast: list->get(-2)
+# The ->get method also works for object keys.
+aValue: $->echo({ a: 123 })->get("a")
+slice: list->slice(0, 5)
+arraySize: array->size
+stringLength: string->size
+numberOfProperties: object->size
+
+# Logical methods
+negation: $.condition->not
+bangBang: $.condition->not->not
+disjunction: $.a->or($.b)->or($.c)
+conjunction: $.a->and($.b, $.c)
+aImpliesB: $.a->not->or($.b)
+excludedMiddle: $.toBe->or($.toBe->not)->eq(true)
```

Comment:
Are we inheriting truthiness rules from any other language? Or is it an error to call these with non-bool expressions?

Reply:
I'm inclined to allow truthiness but always produce booleans from these methods, so `$.condition->and($.object)` doesn't give you the type of `$.object` when `$.condition` is truthy.

Reply:
I think that we should avoid truthiness if possible. It'll probably confuse users more than help. Especially if the user did not mean to use an object as a boolean (e.g. a typo). With implicit truthiness, they might not ever realize.

Reply:
I'm imagining a pretty narrow definition of truthiness: the set of falsy values would include `false`, `null`, `0`, and possibly undefined/missing values, and everything else would be truthy. This would mean the empty string `""` is truthy, unlike in JavaScript, and `[]` or `{}` would both be truthy (unlike in Python).

Reply:
_If_ we decide to have truthiness, do we really want to have our own flavor of it that deviates from what users might be used to in regards to other languages? If anything, I think this shows why being more verbose here with explicit truthiness might be less confusing overall.

Reply:
Alright, let's keep discussing these rules, but I don't think the discussion needs to hold up this PR because we're not shipping the logical methods yet (see https://github.com/apollographql/router/pull/5762/commits/b8cf701648be3b18ec25bbf8b39cb778dd25bf0c).